CONTRIBUTING.md

```diff
@@ -3,11 +3,201 @@ Guidelines for Contributing
+- Avoid to use `auto` as return type wherever possible (makes the documentation easier to read)
+- Use `static` nested structs for ranges (aka [Voldemort types](http://wiki.dlang.org/Voldemort_types)) and thus store function parameters explicitly in the nested struct
+
+### Code advice
+
+- Don't be afraid to use the new compiler or language features (e.g. `Foo!Range` vs. `Foo!(Range)`)
+- Be critical when looking at Phobos code (some parts of Phobos are a bit older and have some relics due to lacking functionality of the compiler)
+- [ ] Is there an easier or more performant way to do it?
+- [ ] Does your PR address only one topic? Can it be split up in smaller PRs?
+- [ ] Is your code flexible enough to accommodate more use cases?
+- [ ] Read through your code again - is there any part that is not understandable?
+
+### Tests
+
+- [Autotester](http://wiki.dlang.org/Git_Commit_Tester) will automatically compile the code in your PR and test it on all supported platforms. For your first PR you need approval from any reviewer - ping them in case they forget.
+- [ ] Do all tests pass locally? You can run the tests of
+	- single module with `rdmd -main -unittest module.d`
+	- an entire package with `make -f posix.mak std/algorithm.test`
+	- all tests with `make -f posix.mak unittest` (add `-j NUMBER_OF_CORES` if you have multiple cores)
+	- for Windows have a look at [building Phobos under windows](http://wiki.dlang.org/Starting_as_a_Contributor#Windows_2))
+- [ ] Do your tests cover all cases? (you can check code coverage in the resulting `module.lst` of `rdmd -cov -main -unittest module.d`)
+
+### Review process
+
+- All PRs _must_ pass all tests for all platforms within [autotester](http://wiki.dlang.org/Git_Commit_Tester) before they will be merged
+- Big additions (new modules) should go to `std.experimental` first (after they pass the [review process](http://wiki.dlang.org/Review/Process))
+- Smaller additions like individual functions can be merged directly after [@andralex](https://github.com/andralex) approves
+- Refactoring or bug fixes just need approval (LGTM) from two reviewers and enough time passed (usually 2-3 days, depending on weekends and holidays) to give everyone an opportunity to shout
+- Trivial changes (e.g. small documentation enhancements) are usually merged in less than 72h - otherwise ping for reviewers
+- If your PR is stalled (no activity for 2-3 days) and you addressed all concerns, don't hesitate to ping your reviewers
+- Ping [@9il](https://github.com/9il) for anything math related
+- See [rebasing](#rebasing) if you need to sync your fork with master
+
+### Naming
+
+- [ ] Do you use clear variable names?
+- [ ] Does your PR have a clear and informative title?
+- [ ] Are your Git messages clear and informative (otherwise see `git rebase`)?
```

Comment:
Why `git rebase`? `git commit --amend` is more appropriate for just changing the commit message.